[Clang] Introduce malloc_span attribute

a-nogikh opened the pull request:
The "malloc" attribute restricts the possible function signatures to the ones returning a pointer, which is not the case for some non-standard allocation function variants. For example, P0901R11 proposed ::operator new overloads that return a return_size_t result - a struct that contains a pointer to the allocated memory as well as the actual size of the allocated memory. Another example is __size_returning_new.

Introduce a new "malloc_span" attribute that exhibits similar semantics, but applies to functions returning records where one member is a pointer (assumed to point to the allocated memory) and another is an integer (assumed to be the size of the allocated memory). This is the case for return_size_t as well as std::span, should it be returned from such an annotated function.

An alternative approach would be to relax the restrictions of the existing "malloc" attribute to be applied to both functions returning pointers and functions returning span-like structs. However, it would complicate the user-space code by requiring specific Clang version checks. In contrast, the presence of a new attribute can be straightforwardly verified via the __has_attribute macro. Introducing a new attribute also avoids concerns about the potential incompatibility with GCC's "malloc" semantics.

In future commits, codegen can be improved to recognize the noalias-ness of the pointer returned inside a span-like struct.

This change helps unlock the alloc token instrumentation for such non-standard allocation functions:
https://clang.llvm.org/docs/AllocToken.html#instrumenting-non-standard-allocation-functions

a-nogikh committed:
[Clang] Introduce malloc_span attribute

The "malloc" attribute restricts the possible function signatures to
the ones returning a pointer, which is not the case for some non-standard
allocation function variants. For example, P0901R11 proposed ::operator new
overloads that return a return_size_t result - a struct that contains
a pointer to the allocated memory as well as the actual size of the
allocated memory. Another example is __size_returning_new.

Introduce a new "malloc_span" attribute that exhibits similar semantics,
but applies to functions returning records whose first member is
a pointer (assumed to point to the allocated memory). This is the case for
return_size_t as well as std::span, should it be returned from such
an annotated function.

An alternative approach would be to relax the restrictions of the
existing "malloc" attribute to be applied to both functions returning
pointers and functions returning span-like structs. However, it would
complicate the user-space code by requiring specific Clang version
checks. In contrast, the presence of a new attribute can be
straightforwardly verified via the __has_attribute macro. Introducing
a new attribute also avoids concerns about the potential incompatibility
with GCC's "malloc" semantics.

In future commits, codegen can be improved to recognize the
noalias-ness of the pointer returned inside a span-like struct.

This change helps unlock the alloc token instrumentation for such
non-standard allocation functions:
https://clang.llvm.org/docs/AllocToken.html#instrumenting-non-standard-allocation-functions

melver commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -6642,7 +6642,8 @@ RValue CodeGenFunction::EmitCall(QualType CalleeType,
     if (CalleeDecl->hasAttr<RestrictAttr>() ||
-        CalleeDecl->hasAttr<AllocSizeAttr>()) {
+        CalleeDecl->hasAttr<AllocSizeAttr>() ||
+        CalleeDecl->hasAttr<MallocSpanAttr>()) {
```
Can we do this in a follow-up PR where we also update the test in the same PR?

melver commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -1839,6 +1839,39 @@ static void handleRestrictAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
 
+static bool isSpanLikeType(const QualType &Ty) {
+  // Check that the type is a plain record with the first field being a pointer
+  // type and the second field being an integer.
```
```suggestion
  // type and the second field being an integer. This matches the  ... <continue paragraph>
```

melver commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -1839,6 +1839,39 @@ static void handleRestrictAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+    return false;
+  const FieldDecl *FirstField = *FieldsBegin;
+  const FieldDecl *SecondField = *std::next(FieldsBegin);
+  return FirstField->getType()->isAnyPointerType() &&
```
It would be easy to allow both orderings (ptr+size, size+ptr). 

For the `malloc` attribute I think we did it for simplicity and not expand the scope too much, but since we're introducing a new attribute entirely, maybe we can permit it right away?

Is there any technical reason we should not allow this now? Because I fear in future someone may want this, and then we're stuck again with version checks.

melver commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -3449,6 +3449,10 @@ def err_attribute_integers_only : Error<
   "%0 attribute only applies to return values that are pointers">,
   InGroup<IgnoredAttributes>;
+def warn_attribute_return_span_only
+    : Warning<"%0 attribute only applies to return values that are span-like "
```
nit: can this be formatted similar to surrounding defs?

a-nogikh committed:
fixup: support different span struct field orderings

Do not demand that the pointer should be the first field and the integer
type the second field.

erichkeane commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -341,6 +327,10 @@ Attribute Changes in Clang
 - New format attributes ``gnu_printf``, ``gnu_scanf``, ``gnu_strftime`` and ``gnu_strfmon`` are added
   as aliases for ``printf``, ``scanf``, ``strftime`` and ``strfmon``. (#GH16219)
 
+- New function attribute `malloc_span` is added. Its semantics is similar to that of the `malloc`
```
```suggestion
- New function attribute `malloc_span` is added. It has semantics similar to that of the `malloc`
```

erichkeane commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -1839,6 +1839,41 @@ static void handleRestrictAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
                  RestrictAttr(S.Context, AL, DeallocE, DeallocPtrIdx));
 }
 
+static bool isSpanLikeType(const QualType &Ty) {
```
Rather than this, what about a `checkSpanLikeType` where we actually do diagnostics?

erichkeane commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -1839,6 +1839,41 @@ static void handleRestrictAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+  const QualType SecondFieldType = std::next(FieldsBegin)->getType();
+  // Verify two possible orderings.
+  return (FirstFieldType->isAnyPointerType() &&
+          SecondFieldType->isIntegerType()) ||
```
I think we should be stricter on integer type.  Also, are we sure we mean `isAnyPointerType`?  This includes some odd ObjC types that I'm not sure we mean.  We should also do some work here (perhaps just writing tests) to ensure that we aren't allowing function member/data member pointers.

I might suggest we also do some work to disallow function pointers unless there is a REALLY good reason not to.

erichkeane replied:
I would do something more restrictive.  Probably an `isSameType` as `Context::PtrDiffT`.

Additionally, I don't see anything in here that prevents these from being bitfields!  That seems like it would be wrong too.

erichkeane commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -1839,6 +1839,41 @@ static void handleRestrictAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+}
+
+static void handleMallocSpanAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+  QualType ResultType = getFunctionOrMethodResultType(D);
```
This doesn't seem to handle dependent result types.  We should have tests that validate that it would work correctly to have the return type be dependent or deduced.

erichkeane replied:
Dependent types is much more important, I suspect 'this doesn't work on deduced return types' is probably acceptable.  

That said, do we work with trailing return types?  Something like:

```
__attribute__((malloc_span)) auto custom_malloc(size_t size)  -> myStruct {
    return (void *)NULL;
}
```
OR a template version of that? We probably SHOULD make sure we get that right.

a-nogikh committed:
fixup: minor update to release notes

a-nogikh committed:
fixup: expand the diagnostic message

a-nogikh committed:
fixup: isSpanLikeType -> checkSpanLikeType

a-nogikh commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -3449,6 +3449,10 @@ def err_attribute_integers_only : Error<
 def warn_attribute_return_pointers_only : Warning<
   "%0 attribute only applies to return values that are pointers">,
   InGroup<IgnoredAttributes>;
+def warn_attribute_return_span_only
```
Expanded the description

erichkeane replied:
This isn't sufficient.  We shouldn't have the warnings contain this much information, it was reasonably fine before.  What we DO need to do however is improve the emission of notes to give 'reasons'.  See all of the `def note_` diagnostics we have.

a-nogikh replied:
Added notes with individual reasons. The validation code has definitely gotten more involved..

a-nogikh committed:
fixup: more restrictions on the pointer type

Disallow function pointers.
Add tests to verify that member/data member pointers also fail.

a-nogikh committed:
fixup: rework malloc_span validation and error reporting

AaronBallman commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -5247,6 +5247,21 @@ yet implemented in clang.
+accessible to the caller.
+
+In this context, a span-like structure is assumed to have two fields, one of
+which is a pointer to the allocated memory and another one is an integer type
```
Yeah, I don't think we need to support smart pointers right away; we can wait for someone to ask for it with a real world use case. As for integers, I think anything >= the width of an int is fine, though `int128_t` would be a surprising count for a span. :-D

Do we think we want any special handling for bit-fields? e.g.,
```
struct S {
  void *ptr;
  int size : 10;
};
```
My gut feeling is: don't diagnose it, if someone does that, there's probably a *reason* for it.

AaronBallman replied:
I could get behind rejecting bit-fields as well until there's a real world use case.

erichkeane commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -3449,6 +3449,18 @@ def err_attribute_integers_only : Error<
+    : Warning<"%0 attribute only applies to functions that return span-like "
+              "structures">,
+      InGroup<IgnoredAttributes>;
+def note_span_must_be_struct : Note<"span-like type must be a struct">;
```
MUST it be a 'struct'? we probably can support class as well, right? 

Also, typically we phrase these in the reverse, something like: `returned type is not a struct/class type`

erichkeane commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -3449,6 +3449,18 @@ def err_attribute_integers_only : Error<
+              "structures">,
+      InGroup<IgnoredAttributes>;
+def note_span_must_be_struct : Note<"span-like type must be a struct">;
+def note_span_must_be_complete : Note<"span-like type must be complete">;
```
`returned type is incomplete`/etc.

a-nogikh committed:
fixup: update documentation

Reflect that pointer+pointer is also possible now.

a-nogikh committed:
fixup: refactor semantic checks

Change error descriptions and restructure the checks code.

melver commented on clang/test/Sema/attr-malloc_span.c:
```diff
@@ -0,0 +1,83 @@
+// RUN: %clang_cc1 -verify -fsyntax-only %s
+// RUN: %clang_cc1 -emit-llvm -o %t %s
```
Why the additional `-emit-llvm` case?
I know the attr-malloc.c test has it, but we're not testing CodeGen in Sema tests.

a-nogikh committed:
[experimental] Support templates

Support returned dependent types and class tempate instantiations.

a-nogikh committed:
fixup: remove typos

Fix typos in docs and diagnostic messages.
Remove unnecessary line in a test file.

erichkeane commented on clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:
```diff
@@ -796,6 +796,20 @@ static void instantiateDependentHLSLParamModifierAttr(
+                                               Decl *New) {
+  QualType RT = getFunctionOrMethodResultType(New);
+  if (RT->isDependentType()) {
+    // The type is still dependent.
```
Same regarding comments & curleys, comments aren't adding anything.  AND if the dependence check is in the check function, this all becomes:

```
if (!CheckSpanLikeType(...))
New->addAttr(...);
```

vastly simplifying & removing repetition

a-nogikh committed:
fixup: add trailing return type tests

a-nogikh committed:
fixup: simplify dependent type processing code

AaronBallman commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -1839,6 +1839,65 @@ static void handleRestrictAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+  if (!RD || RD->isUnion())
+    return emitWarning(diag::note_returned_not_struct);
+  auto FieldsBegin = RD->field_begin();
+  const auto FieldsCount = std::distance(FieldsBegin, RD->field_end());
```
Just making sure we're okay with this, but I think this means we're accepting:
```
struct S {
  void *ptr;
  int len;
  static int not_a_field;
};
```
(We should probably have a test case showing we thought about this situation. I don't have strong opinions on accept vs reject for it, but lean towards reject.)

Also, what do we think about base classes? e.g.,
```
struct S {
  void *ptr;
};

struct D : S {
  void *ptr;
  int len;
};
```
I don't think that's span-like, so I think the presence of a non-empty base class should probably be an error?

Also, per our coding standards, you should drop the top-level `const` uses from this function (it's fine to have `const T *` but `const T` is something we generally don't do for local variables).

a-nogikh replied:
I've pushed the fix commits that:
* Remove `const T` for local variables.
* Add a test for structs with static fields.
* Prohibit non-empty base classes.

a-nogikh committed:
fixup: fix typos

a-nogikh committed:
fixup: add a test for static struct field members

Ensure static fields are ignored.

a-nogikh committed:
fixup: drop const for non-pointers

Conform to the coding style.

a-nogikh committed:
fixup: verify span-like struct/class bases

If present, they must be empty.

AaronBallman commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -3449,6 +3449,24 @@ def err_attribute_integers_only : Error<
+    : Note<"integer field #%0 of span-like type is not wide enough (minimum "
+           "width: %1)">;
+def note_inherits_not_empty_base
+    : Note<"returned struct/class inherits from a non-empty base">;
```
This one could use `class` instead of `struct` because inheritance is a C++ thing, or we could look at which tag is used in practice and print that with `%select`.

erichkeane reviewed (commented):
1 suggestion on base classes, else I'm OK.  Make Aaron happy, and I am.

a-nogikh committed:
fixup: improve the docs and diagnostic messages

a-nogikh committed:
fixup: reject types inheriting from any base

a-nogikh committed:
fixup: add a virtual base class semantic test

AaronBallman reviewed (approved):
LGTM assuming precommit CI comes back green

a-nogikh commented:
I am not sure there's any way to guarantee this: it's an allocator implementation detail. Conceptually, yes, we should assume this to be the case, and it's true for the mentioned use cases (e.g. the size returning operator new, which allocates at least the requested amount of memory).
